Lib/_strptime.py

```diff
@@ -577,3 +582,19 @@ def _strptime_datetime(cls, data_string, format="%a %b %d %H:%M:%S %Y"):
         args += (tz,)
 
     return cls(*args)
+
+def _strptime_datetime_date(data_string, format):
+    """Return a date based on the input string and the format string."""
+    msg = "'{!s}' {} not valid in date format specification."
+    from _datetime import _check_invalid_datetime_specs
+    if _check_invalid_datetime_specs(format, time_specs, msg):
```

Comment:
Why not move the _msg_ string into the _check_ function, and just pass the bit that varies ('date' or 'time') as an argument? It would make the code easier to read.

Looks like the _check_ function either raises an exception or returns True. It would be clearer to not use an _if_ statement here.

Reply:
at the beginning, this PR was a submitted patch by other contributors, I just wanted to convert it to a PR. and now, I try to fix all the issues.

Reply:
but for this point, the first step was the conversion of `_check_invalid_datetime_specs` to C, and after start to improve the code with the other recommendations. (from @vstinner and @pganssle)